### PR TITLE
Ct-581 grants as configs

### LINKED ISSUE #5189: [CT-581] [Feature] Grants as Node Configs
### Description

This is the first in a two ticket series. #5263 is the second ticket. They will be merged sequentially, but both are required for this feature to be exposed to users.

Today users often configure a `post_hook` to grant permissions on models, seeds, and snapshots:
```
{{ config(post_hook = 'grant select on {{ this }} to role reporter') }}
```

These two tickets aim to make it easier to specify grants allowing it to be configured directly both in dbt_project.yml as well as in source files:
```
# dbt_project.yml
models:
  export:
    +grants:
      select: ['reporter', 'bi']
```
```
-- SQL
{{ config(grants = {'select': ['other_user']}) }}
```

These grant configs will not necessarily look the same for each and every warehouse. The logic to generate the sql from these configs can be overriden by adapters.

### Implementation

- [ ] a "grants" config should be added to model, seed, and snapshot nodes. The value of this config is a dictionary.
- [ ] merge behavior for grants should be to override / clobber not append.
- [ ] test that projects can define grants in 1) the dbt_project.yml, 2) models, seeds, and snapshots, and 3) both the dbt_project.yml and models, seeds, and snapshots with the correct merge behavior.

### PR DESCRIPTION
**Pull request by gshank**
resolves #5189

### Description

Add "grants" attribute to NodeConfig. Fix up all the tests.

This also contains a fix for a partial parsing error that was encountered when writing the test.

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR
- [x] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md#Adding-CHANGELOG-Entry)

### CODE REVIEW

**Comment by gshank** [comment]
@jtcohen6 We weren't really clear on whether we want to limit the keys that can be contained in the 'grants' dictionary.  I went forward assuming that people might want to create and support their own particular keys, and also that using something really specific like "select" would feel wrong on some warehouses. Or would you prefer to limit the legal keys in the grant dictionary?

I used "my_select" in the test instead of "select" because the test would probably break when we actually implement a "select" grant macro.

**Comment by gshank** [comment]
I also noticed that the jsonschema artifact test doesn't fail when we add a new attribute to NodeConfig, probably because of the really clever things we do to allow random keys in configs.  I would have updated the jsonschemas, but I'm not sure whether we want to update to a v6 jsonschema now (and keep updating as things change) or do later.

**Comment by gshank** [comment]
The partial parsing error was that if you have config in a SQL file that's been parsed and add a new schema file that also has config, the config from the SQL file will get lost because I wasn't saving the 'config_call_dict' in the serialized manifest. I switched to saving that dictionary and deleting config_call_dict for the manifest.json instead.  Updates to the schema file will work correctly. Any update to either file would correct the error.

This feels like an edge case and possibly not worth doing as a backport, but if you'd like it backported let me know and I can separate it out.

File: tests/functional/configs/test_grant_configs.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+        manifest = get_manifest(project.project_root)
+        model_id = "model.test.my_model"
+        assert model_id in manifest.nodes
+
+        model = manifest.nodes[model_id]
+        model_config = model.config
+        assert hasattr(model_config, "grants")
+
+        expected = {"my_select": ["reporter", "bi", "other_user"]}
```
Comment (jtcohen6) [inline tests/functional/configs/test_grant_configs.py:53]: There needs to be some mechanism to remove grants that have been applied at a broader level of the configuration hierarchy.

So: I think we need `other_user` to clobber the existing values of the `my_select` key (`["reporter", "bi"]`), rather than append to the list of recipients for the `my_select` privilege.

The ideal case is one in which users have some control over this merging behavior. Is now a good time to revisit the discussion in https://github.com/dbt-labs/dbt-core/issues/4108? (Even if we don't / can't actually enable that now)

**Comment by jtcohen6** [comment]
> I went forward assuming that people might want to create and support their own particular keys, and also that using something really specific like "select" would feel wrong on some warehouses. Or would you prefer to limit the legal keys in the grant dictionary?

I agree with your working assumption. I wouldn't want to have to update this every time a warehouse released a new privilege type. It's also infeasible on BigQuery, which supports tons of privileges, and custom privileges.

> I would have updated the jsonschemas, but I'm not sure whether we want to update to a v6 jsonschema now (and keep updating as things change) or do later.

Worth thinking about:
- Pairing this with the short-term salve for users of Slim CI, proposed in https://github.com/dbt-labs/dbt-core/issues/5213#issuecomment-1117301365, for artifact updates that actually have backwards-compatible changes
- Whether we actually need to update the artifact version for non-breaking changes, such as an additional default value in `node.config` (which can already contain any additional properties)

Nice catch on the partial parsing error! I haven't heard of someone running into this yet, so I'm happy to keep it as a forthcoming bug fix for now, without the backport.

**Comment by gshank** [comment]
I changed to default of clobbering the grants with the ability of extending them if the grant key starts with '+'. Take a look and see what you think.

File: core/dbt/contracts/graph/manifest.py [hunk 2]
```diff
@@ -1135,6 +1135,12 @@ class WritableManifest(ArtifactMixin):
         )
     )
 
+    def __post_serialize__(self, dct):
```
[the remark this answers is not in the document]
Comment (gshank) [reply core/dbt/contracts/graph/manifest.py:1138]: The problem we were running into with the partial parsing bug was that the config_call_dict wasn't being saved with the partial parsing manifest. So when a new schema file was added the config was being applied to the existing model node, but without the config_call_dict we lose the config from the SQL file.

There were two options to fix it, 1) save the config_call_dict or 2) every time we add a new schema file reload all of the nodes that *might* be affected. The second option would complicate partial parsing substantially and felt more inconsistent. The config_call_dict wasn't being saved to start with only because my imagination did not think about this case.  I clean it up in the WritableManifest post_serialize method because we don't really want or need it in the manifest artifact.

**Comment by gshank** [comment]
@ChenyuLInx Your sum up is correct.

File: core/dbt/context/context_config.py [hunk 3]
```diff
@@ -264,18 +264,49 @@ def add_config_call(self, opts: Dict[str, Any]) -> None:
...
             if k in BaseConfig.mergebehavior["append"]:
                 if not isinstance(v, list):
                     v = [v]
-            if k in BaseConfig.mergebehavior["update"] and not isinstance(v, dict):
-                raise InternalException(f"expected dict, got {v}")
-            if k in config_call_dict and isinstance(config_call_dict[k], list):
-                config_call_dict[k].extend(v)
-            elif k in config_call_dict and isinstance(config_call_dict[k], dict):
-                config_call_dict[k].update(v)
+                if k in config_call_dict and isinstance(config_call_dict[k], list):
```
[the remark this answers is not in the document]
Comment (gshank) [reply core/dbt/context/context_config.py]: I don't think it would cause any problems, but you're right that it's unnecessary. I've removed the isinstance check.

File: core/dbt/context/context_config.py [hunk 4]
```diff
@@ -264,18 +264,49 @@ def add_config_call(self, opts: Dict[str, Any]) -> None:
...
+                if not isinstance(v, dict):
+                    raise InternalException(f"expected dict, got {v}")
+                if k in config_call_dict and isinstance(config_call_dict[k], dict):
+                    config_call_dict[k].update(v)
+                else:
+                    config_call_dict[k] = v
+            elif k in BaseConfig.mergebehavior["dict_key_append"]:
+                if not isinstance(v, dict):
+                    raise InternalException(f"expected dict, got {v}")
+                if k in config_call_dict and isinstance(config_call_dict[k], dict):
```
[the remark this answers is not in the document]
Comment (gshank) [reply core/dbt/context/context_config.py]: I've removed the isinstance check. It ought to be unnecessary.

**Review by ChenyuLInx — APPROVED** [review]
LGTM!
